File: course/section_5_binary_search/code/12-valid-perfect-square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        left, right = 0, num
        while left <= right:
            m = (left + right) // 2
            sqr = m**2
            if sqr == num:
                return True
            if sqr > num:
                right = m - 1
            else:
                left = m + 1
        return False

File: course/section_5_binary_search/code/12-valid-perfect-square.php
<?php

class Solution {
    function isPerfectSquare($num) {
        $left = 0 ;
        $right = $num ; 
        while($left <= $right){
            $m = intdiv($left + $right , 2) ;
            $square = $m * $m ;
            if($square=== $num){
                return True; 
            } elseif( $square > $num){
                $right = $m - 1; 
            } else {
                $left = $m + 1 ;
            }
        }
        return False ;
    }
}